src/bosh_azure_cpi/lib/cloud/azure/vms/instance_type_mapper.rb

```diff
@@ -4,95 +4,88 @@ module Bosh::AzureCloud
-    # 3. General purpose (Balanced CPU-to-memory ratio) > Compute optimized (High CPU-to-memory ratio) > Memory optimized (High memory-to-CPU ratio);
-    # 4. High performance compute is listed at the end;
-    # 4. The following sizes are not recommended:
-    #      a. The sizes with GPU (NV, NC, ND);
-    #      b. The M-series which has lots of CPUs and memories.
-    #      c. The Basic tier sizes (they don't support load balancing);
-    RECOMMENDED_VM_SIZES = [
-      # Premium Storage
-      %w[Standard_D2s_v3 Standard_D4s_v3 Standard_D8s_v3 Standard_D16s_v3 Standard_D32s_v3 Standard_D64s_v3],
-      %w[Standard_E2s_v3 Standard_E4s_v3 Standard_E8s_v3 Standard_E16s_v3 Standard_E32s_v3 Standard_E64s_v3 Standard_E64is_v3],
-      %w[Standard_DS1_v2 Standard_DS2_v2 Standard_DS3_v2 Standard_DS4_v2 Standard_DS5_v2],
-      %w[Standard_F2s_v2 Standard_F4s_v2 Standard_F8s_v2 Standard_F16s_v2 Standard_F32s_v2 Standard_F64s_v2 Standard_F72s_v2],
-      %w[Standard_DS11_v2 Standard_DS12_v2 Standard_DS13_v2 Standard_DS14_v2 Standard_DS15_v2],
-      %w[Standard_B1s Standard_B1ms Standard_B2s Standard_B2ms Standard_B4ms Standard_B8ms],
-      %w[Standard_F1s Standard_F2s Standard_F4s Standard_F8s Standard_F16s],
-      %w[Standard_GS1 Standard_GS2 Standard_GS3 Standard_GS4 Standard_GS5],
-      %w[Standard_DS1 Standard_DS2 Standard_DS3 Standard_DS4 Standard_DS11 Standard_DS12 Standard_DS13 Standard_DS14],
-      # Standard Storage
-      %w[Standard_D2_v3 Standard_D4_v3 Standard_D8_v3 Standard_D16_v3 Standard_D32_v3 Standard_D64_v3],
-      %w[Standard_E2_v3 Standard_E4_v3 Standard_E8_v3 Standard_E16_v3 Standard_E32_v3 Standard_E64_v3 Standard_E64i_v3],
-      %w[Standard_D1_v2 Standard_D2_v2 Standard_D3_v2 Standard_D4_v2 Standard_D5_v2],
-      %w[Standard_A1_v2 Standard_A2_v2 Standard_A4_v2 Standard_A8_v2 Standard_A2m_v2 Standard_A4m_v2 Standard_A8m_v2],
-      %w[Standard_D11_v2 Standard_D12_v2 Standard_D13_v2 Standard_D14_v2 Standard_D15_v2],
-      %w[Standard_F1 Standard_F2 Standard_F4 Standard_F8 Standard_F16],
-      %w[Standard_G1 Standard_G2 Standard_G3 Standard_G4 Standard_G5],
-      %w[Standard_D1 Standard_D2 Standard_D3 Standard_D4 Standard_D11 Standard_D12 Standard_D13 Standard_D14],
-      %w[Standard_A1 Standard_A2 Standard_A3 Standard_A4 Standard_A5 Standard_A6 Standard_A7 Standard_A8 Standard_A9 Standard_A10 Standard_A11],
-      # High performance compute
-      %w[Standard_H8 Standard_H16 Standard_H8m Standard_H16m Standard_H16r Standard_H16mr]
-    ].freeze
-
-    def map(desired_instance_size, available_vm_sizes)
+    SERIES_PREFERENCE = {
+      'D' => 5,  # General purpose - balanced
+      'F' => 4,  # Compute optimized
+      'E' => 4,  # Memory optimized
+      'B' => 3,  # Burstable VMs
+      'L' => 3,  # Storage optimized
+      'A' => 2,  # Basic general purpose
+      'H' => 1,  # High performance compute
+      'M' => 1,  # Ultra memory optimized
+      'N' => 1   # GPU
```

Comment:
M and N series VMs were explicitly excluded previously, likely due to underlying cost. While they are at the bottom of the preference list, do we think it's safe to automatically choose them based on `vm_resources` key?

Reply:
I do somewhat have a preference for not adding the more expensive VM types. Their presence could make typos in the manifest (accidentally add an extra 0 after RAM value) launch an unexpectedly expensive VM. It feels like if you need something that high end or special purpose like the GPU instances, you should opt-in to it explicitly via cloud config.

(By the same logic I don't think things like the H series VMs belong in this list, but they were already there, so I'm a little more reluctant to remove them)